src/pivot-table/hooks/use-column-width.ts

```diff
@@ -1,167 +1,245 @@
     },
   } = layoutService;
   const styleService = useStyleContext();
+  const { measureText: measureTextForHeader } = useMeasureText(styleService.header);
+  const { estimateWidth: estimateWidthForRowContent } = useMeasureText(styleService.rowContent);
+  const { measureText: measureTextForColumnContent, estimateWidth: estimateWidthForColumnContent } = useMeasureText(
+    styleService.columnContent,
+  );
   const { estimateWidth: estimateWidthForContent, measureText: measureTextForContent } = useMeasureText(
     styleService.content,
   );
-  const { measureText: measureTextForHeader } = useMeasureText(styleService.header);
 
-  const hasPseudoDimOnLeft = useMemo(() => visibleLeftDimensionInfo.includes(-1), [visibleLeftDimensionInfo]);
+  const leafTopDimension = visibleTopDimensionInfo.at(-1);
+  const topGridLeavesIsPseudo = leafTopDimension === PSEUDO_DIMENSION_INDEX;
 
-  const leftColumnWidthsRatios = useMemo(() => {
-    const ratios = visibleLeftDimensionInfo.map((qDimensionInfo, index) => {
-      if (qDimensionInfo === PSEUDO_DIMENSION_INDEX) {
-        const pseudoDimensionWidth = Math.max(...qMeasureInfo.map((m) => measureTextForContent(m.qFallbackTitle)));
+  const getCollapseExpandIconSize = useCallback(
+    (index: number) => (index < qNoOfLeftDims - 1 ? EXPAND_ICON_WIDTH : 0),
+    [qNoOfLeftDims],
+  );
 
-        return pseudoDimensionWidth / rect.width;
+  /**
+   * The widths of the left columns. Scales the width to fit LEFT_SIDE_MAX_WIDTH_RATIO * rect.width if wider than that
+   */
+  const leftGridColumnWidths = useMemo(() => {
+    const getColumnWidth = (columnWidth: ColumnWidth, fitToContentWidth: number) => {
+      switch (columnWidth?.type) {
+        case ColumnWidthType.Pixels:
+          return getPixelValue(columnWidth.pixels);
+        case ColumnWidthType.Percentage:
+          return getPercentageValue(columnWidth.percentage) * rect.width;
+        default:
+          // fit to content / auto
+          return fitToContentWidth;
       }
+    };
 
-      const { qFallbackTitle, qApprMaxGlyphCount } = qDimensionInfo;
-      const hasChildNodes = index < qNoOfLeftDims - 1; // -1 as the last column can not be expanded or collapsed
-      const collapseExpandIconSize = hasChildNodes ? EXPAND_ICON_WIDTH : 0;
-      const w = Math.max(
-        measureTextForHeader(qFallbackTitle),
-        estimateWidthForContent(qApprMaxGlyphCount) + collapseExpandIconSize,
-      );
-      return w / rect.width;
+    let sumOfWidths = 0;
+
+    const widths = visibleLeftDimensionInfo.map((qDimensionInfo, index) => {
+      let width;
+
+      if (qDimensionInfo === PSEUDO_DIMENSION_INDEX) {
+        // Use the max width of all measures
+        width = Math.max(
+          ...qMeasureInfo.map(({ qFallbackTitle, columnWidth }) => {
+            const fitToContentWidth = measureTextForContent(qFallbackTitle);
+            return getColumnWidth(columnWidth, fitToContentWidth);
+          }),
```

Comment:
I think this can be externalised from it's outer map which happens on line 90 `visibleLeftDimensionInfo.map(...)`.
currently it's a small O(n*2) operation happens, I expect it's not that much heavy operation for all executions since it's only happens for pseudo dims. but it's good to externalise it since what it requires is only `qMeasureInfo` which `visibleLeftDimensionInfo` is not coupled with it at all.

Reply:
Well, there can only be one pseudo dimension in `visibleLeftDimensionInfo`, so `qMeasureInfo.map` can only run once